From b5d5f7fa3107e5404d5bac24fd7093ba4f0440be Mon Sep 17 00:00:00 2001
From: Rokis <ghost1988102@gmail.com>
Date: Sun, 5 Jul 2020 08:26:33 +0800
Subject: [PATCH 057/122] usb2jffs: 1.Add USB2JFFS function(include web,
 scripts and settings)           2.Add action_mode "toolscript" to exec script
 in asp.           3.Add formatting "EXT4" into disk_format.

Signed-off-by: Rokis <ghost1988102@gmail.com>
---
 release/src/router/httpd/sysinfo.c            |  15 +
 release/src/router/httpd/web.c                |   8 +
 release/src/router/others/Makefile            |   2 +
 release/src/router/others/usb2jffs.sh         | 252 ++++++++++
 release/src/router/rc/usb.c                   |  19 +-
 release/src/router/shared/defaults.c          |   3 +
 release/src/router/www/Tools_Usb2Jffs.asp     | 445 ++++++++++++++++++
 release/src/router/www/ajax_sysinfo.asp       |   2 +
 .../src/router/www/device-map/disk_format.asp |  14 +-
 .../menuTree_bwdpi_traffic_analyzer.js        |   1 +
 .../require/menuTrees/menuTree_no_bwdpi.js    |   1 +
 11 files changed, 760 insertions(+), 2 deletions(-)
 create mode 100755 release/src/router/others/usb2jffs.sh
 create mode 100644 release/src/router/www/Tools_Usb2Jffs.asp

diff --git a/release/src/router/httpd/sysinfo.c b/release/src/router/httpd/sysinfo.c
index b6ee62fcf7..0d06433806 100644
--- a/release/src/router/httpd/sysinfo.c
+++ b/release/src/router/httpd/sysinfo.c
@@ -509,6 +509,21 @@ int ej_show_sysinfo(int eid, webs_t wp, int argc, char_t ** argv)
 #else	// HND lacks robocfg support
 			strcpy(result, "[]");
 #endif
+		} else if( strcmp(type,"usb2jffs.mount") == 0 ) {
+			FILE *fp;
+			char device[30], path[30];
+			sprintf(result, "%d", 0);
+			fp = fopen("/proc/mounts", "r");
+			if (fp) {
+				while(!feof(fp))  {
+					fscanf(fp, "%s %s", device, path);
+					if (strcmp(path, "/jffs") == 0 && strstr(device, "/dev/s")) {
+						sprintf(result, "%d", 1);
+						break;
+					}
+				}
+				fclose(fp);
+			}
 		} else if(strlen(type) > 8 && strncmp(type,"hwaccel", 7) == 0 ) {
 			if (!strcmp(&type[8], "runner"))	// Also query Archer on 675x
 #if defined(RTAC86U) || defined(GTAC2900)
diff --git a/release/src/router/httpd/web.c b/release/src/router/httpd/web.c
index 04640f498f..d10c841c48 100644
--- a/release/src/router/httpd/web.c
+++ b/release/src/router/httpd/web.c
@@ -5246,6 +5246,14 @@ static int ej_update_variables(int eid, webs_t wp, int argc, char_t **argv)
 		notify_rc(notify_cmd);
 	}
 #endif
+	else if(!strcmp(action_mode, "toolscript")){
+		nvram_set("freeze_duck", "5");
+		strncpy(notify_cmd, action_script, 128);
+		_dprintf("Scrip_Cmd: [%s]\n", notify_cmd);
+		validate_apply(wp, NULL);
+		strlcpy(SystemCmd, notify_cmd, sizeof(SystemCmd));
+		sys_script("syscmd.sh");
+	}
 
 	return 0;
 }
diff --git a/release/src/router/others/Makefile b/release/src/router/others/Makefile
index 67319703f0..9e089c9673 100755
--- a/release/src/router/others/Makefile
+++ b/release/src/router/others/Makefile
@@ -11,6 +11,8 @@ install:
 	install -m 0700 wlk $(INSTALLDIR)/usr/sbin
 # Gtoolbox
 	install -m 0755 gtoolbox $(INSTALLDIR)/usr/sbin
+# usb2jffs
+	install -m 0755 usb2jffs.sh $(INSTALLDIR)/usr/sbin
 
 ifeq ($(RTCONFIG_USB),y)
 # Optware perl symlink
diff --git a/release/src/router/others/usb2jffs.sh b/release/src/router/others/usb2jffs.sh
new file mode 100755
index 0000000000..c907e62d76
--- /dev/null
+++ b/release/src/router/others/usb2jffs.sh
@@ -0,0 +1,252 @@
+#!/bin/sh
+
+usb_type="ext3|ext4|thfsplus"
+usb_path=$(mount | grep "/tmp/mnt/" | grep -E ${usb_type} | grep "/dev/s" | head -1 | awk '{print $3}')
+usb_folder="${usb_path}/.usb_jffs"
+
+stop_software_center(){
+	killall skipd >/dev/null 2>&1
+	/koolshare/perp/perp.sh stop >/dev/null 2>&1
+}
+
+start_software_center(){
+	killall skipd >/dev/null 2>&1
+	/koolshare/perp/perp.sh stop >/dev/null 2>&1
+	service start_skipd >/dev/null 2>&1
+	/koolshare/perp/perp.sh start >/dev/null 2>&1
+}
+
+mount_jffs(){
+	if [ -z "$usb_path" ]; then
+		echo "没有找到符合要求的USB存储器！"
+		return
+	fi
+	sync
+	if [ ! -d "$usb_folder" ]; then
+		echo "没有找到${usb_folder}！"
+		echo "复制原始分区/jffs内容到${usb_folder}！"
+		mkdir -p "$usb_folder"
+		cp -af /jffs/. "$usb_folder"
+	else
+		echo "找到${usb_folder}！"
+		echo "以前挂载过USB型JFFS分区，将为你重新挂载！"
+	fi
+	echo "停止软件中心相关进程！"
+	stop_software_center
+	echo "卸载JFFS分区..."
+	umount -l /jffs
+	echo "将${usb_folder}挂载到JFFS分区..."
+	mount -o rbind "$usb_folder" /jffs
+	if [ "$?" == "0" ]; then
+		echo "USB型JFFS分区挂载成功！"
+		echo "将ubi0_0挂载在/cifs2"
+		mount -t ubifs -o rw,noatime ubi0_0 /cifs2
+		if [ "$?" == "0" ]; then
+			echo "/cifs2挂载成功！"
+		else
+			echo "/cifs2挂载失败，将无法使用同步功能！！"
+		fi
+	else
+		echo "USB型JFFS挂载失败！！"
+		echo "恢复原始挂载分区！"
+		mount -t ubifs -o rw,noatime ubi0_0 /jffs
+		if [ "$?" == "0" ]; then
+			echo "已经恢复原始挂载分区！"
+		else
+			echo "恢复原始挂载分区失败，请重启路由器后重试！"
+		fi
+	fi
+	echo "重启软件中心相关进程..."
+	start_software_center
+	echo "完成！"
+}
+
+unmount_jffs(){
+	usb2jffs_rsync=`nvram get usb2jffs_rsync`
+	if [ "$usb2jffs_rsync" -eq "1" ]; then
+		echo "检测到卸载时同步已开启，开始同步！"
+		sync_usb_mtd
+	fi
+	echo "开始卸载！"
+	sync
+	echo "停止软件中心相关进程！"
+	stop_software_center
+	echo "卸载/cifs2..."
+	umount -l /cifs2
+	if [ "$?" == "0" ]; then
+		echo "/cifs2卸载成功..."
+	else
+		echo "/cifs2卸载失败！"
+	fi
+	echo "卸载USB型JFFS..."
+	umount -l /jffs
+	if [ "$?" == "0" ]; then
+		echo "USB型JFFS卸载成功..."
+		echo "将文件系统ubi0_0挂载到jffs分区..."
+		mount -t ubifs -o rw,noatime ubi0_0 /jffs
+		if [ "$?" == "0" ]; then
+			echo "原始分区挂载成功！"
+			echo "重启软件中心相关进程..."
+			start_software_center
+		else
+			echo "原始分区分区挂载失败！请重启后再试！"
+		fi
+	else
+		echo "卸载USB型JFFS分区挂载失败！请重启后再试！"
+	fi
+}
+
+remove_jffs(){
+	unmount_jffs
+	echo "删除${usb_folder}！"
+	rm -rf "$usb_folder"
+}
+
+sync_usb_mtd(){
+	# 检测是否有USB磁盘挂载在/jffs上
+	local mounted_nu=$(mount | grep "/jffs" | grep -c "/dev/s")
+	if [ "$mounted_nu" -ge "1" ]; then
+		echo "检测到USB磁盘的${usb_folder}挂载在/jffs上"
+	else
+		echo "没有检测到任何挂载在/jffs上的USB磁盘设备，退出同步！"
+		return
+	fi
+
+	# 检查cifs2是否挂载
+	local cifs2_device=$(df -h | grep cifs2 | awk '{print $1}' 2>&1)
+	if [ -z "${cifs2_device}" -o "${cifs2_device}" != "ubi0_0" ]; then
+		echo "/cifs2未挂载！文件同步无法进行，请重启路由器后重试！"
+		return
+	fi
+
+	# 检测jffs分区剩余空间...
+	local TOTAL_MTD_SIZE=$(df | grep cifs2 | awk '{print $2}')
+	local SPACE_USB_JFFS=$(du -s ${usb_folder} | awk '{print $1}')
+	if [ "${TOTAL_MTD_SIZE}" -gt "${SPACE_USB_JFFS}" ];then
+		# echo "空间满足，继续下一步！"
+		# 文件夹同步
+		# ls 找出所有的目录，写入文件
+		ls -aelR /jffs/ | grep -i "^/" | sed 's/:$//g' | sed 's/^\/jffs\///g' | sed '/^$/d' > /tmp/upload/jffs_folders.txt
+		ls -aelR /cifs2/ | grep -i "^/" | sed 's/:$//g' | sed 's/^\/cifs2\///g' | sed '/^$/d' > /tmp/upload/cifs2_folders.txt
+		# 文件同步
+		# find 找出/jffs下所有的文件+目录，ls列出来后，只取文件的名字，写入文件（因为固件不支持-type f，只能这么搞了）
+		find /jffs -exec ls -ael {} \; | sed '/^d/d' | awk '{print $11}' | sed '/^\//!d' | sed 's/^\/jffs\///g' | sed '/^$/d' | sort > /tmp/upload/jffs_files.txt
+		find /cifs2 -exec ls -ael {} \; | sed '/^d/d' | awk '{print $11}' | sed '/^\//!d' | sed 's/^\/cifs2\///g' | sed '/^$/d' | sort > /tmp/upload/cifs2_files.txt
+	else
+		echo "ubi0_0总容量为：${TOTAL_MTD_SIZE}KB, USB磁盘中.usb_jffs文件夹大小：${SPACE_USB_JFFS}KB"
+		echo "ubi0_0空间容量不够，只进行系统相关文件的同步，软件中心和插件文件不进行同步！"
+		# 文件夹同步
+		# ls 找出所有的目录，写入文件
+		ls -aelR /jffs/ | grep -i "^/" | sed 's/:$//g' | sed 's/^\/jffs\///g' | sed '/^$/d' | sed '/koolshare/d' | sed '/dnsmasq.d/d' | sort > /tmp/upload/jffs_folders.txt
+		ls -aelR /cifs2/ | grep -i "^/" | sed 's/:$//g' | sed 's/^\/cifs2\///g' | sed '/^$/d' | sed '/koolshare/d' | sed '/dnsmasq.d/d' | sort > /tmp/upload/cifs2_folders.txt
+		# 文件同步
+		# find 找出/jffs下所有的文件+目录，ls列出来后，只取文件的名字，写入文件（因为固件不支持-type f，只能这么搞了）
+		find /jffs -exec ls -ael {} \; | sed '/^d/d' | awk '{print $11}' | sed '/^\//!d' | sed 's/^\/jffs\///g' | sed '/^$/d' | sed '/koolshare/d' | sed '/dnsmasq.d/d' | sort > /tmp/upload/jffs_files.txt
+		find /cifs2 -exec ls -ael {} \; | sed '/^d/d' | awk '{print $11}' | sed '/^\//!d' | sed 's/^\/cifs2\///g' | sed '/^$/d' | sed '/koolshare/d' | sed '/dnsmasq.d/d' | sort > /tmp/upload/cifs2_files.txt
+	fi
+	# 开始同步
+	echo "文件同步：USB_JFFS → MTD_JFFS！（/jffs → /cifs2）"
+	if [ -s /tmp/upload/cifs2_folders.txt -a -s /tmp/upload/jffs_folders.txt ]; then
+		# /cifs2和/jffs下都有目录的情况下
+		# MTD_JFFS目录下独有的目录
+		local mtd_uniq_folders=$(awk 'NR==FNR{a[$1]=$1} NR>FNR{if(a[$1] == ""){print $1}}' /tmp/upload/jffs_folders.txt /tmp/upload/cifs2_folders.txt)
+		for mtd_uniq_folder in ${mtd_uniq_folders}; do
+			echo "删除文件夹：${mtd_uniq_folder}"
+			rm -rf /cifs2/${mtd_uniq_folder}
+		done
+
+		# USB_JFFS目录下独有的目录
+		local usb_uniq_folders=$(awk 'NR==FNR{a[$1]=$1} NR>FNR{if(a[$1] == ""){print $1}}' /tmp/upload/cifs2_folders.txt /tmp/upload/jffs_folders.txt)
+		for usb_uniq_folder in ${usb_uniq_folders}; do
+			echo "添加文件夹：${usb_uniq_folder}"
+			mkdir -p /cifs2/${usb_uniq_folder}
+		done
+	elif [ ! -s /tmp/upload/cifs2_folders.txt -a -s /tmp/upload/jffs_folders.txt ]; then
+		# CIFS2下没有任何目录，USB_JFFS所有目录都是独有的
+		local usb_uniq_folders=$(cat /tmp/upload/jffs_folders.txt)
+		for usb_uniq_folder in ${usb_uniq_folders}; do
+			echo "添加文件夹：${usb_uniq_folder}"
+			mkdir -p /cifs2/${usb_uniq_folder}
+		done
+	else
+		echo "文件同步：未知错误，请重启或者格式化USB磁盘后重试！"
+		return
+	fi
+
+	# MTD_JFFS目录下独有的文件
+	local mtd_uniq_files=$(awk 'NR==FNR{a[$1]=$1} NR>FNR{if(a[$1] == ""){print $1}}' /tmp/upload/jffs_files.txt /tmp/upload/cifs2_files.txt)
+	for mtd_uniq_file in ${mtd_uniq_files}; do
+		echo "删除文件：${mtd_uniq_file}"
+		rm -rf /cifs2/${mtd_uniq_file}
+	done
+
+	# USB_JFFS目录下独有的文件
+	local usb_uniq_files=$(awk 'NR==FNR{a[$1]=$1} NR>FNR{if(a[$1] == ""){print $1}}' /tmp/upload/cifs2_files.txt /tmp/upload/jffs_files.txt)
+	for usb_uniq_file in ${usb_uniq_files}; do
+		echo "添加文件：${usb_uniq_file}"
+		cp -af /jffs/${usb_uniq_file} /cifs2/${usb_uniq_file}
+	done
+
+	# 更新变化的文件
+	# 1. × 使用时间戳，在生成文件列表的时候附带生成时间戳，用以判断文件是否变化，但是实际测试但是很不准，但是可以告知用户哪些文件变化了
+	# 2. √ 使用md5sum，每个文件都要比较，非常消耗！！但是这个至少比较准确，可以告知用户那些文件变化了
+	local FILES=$(cat /tmp/upload/jffs_files.txt)
+	for FILE in ${FILES}
+	do
+		USB_MD5=$(md5sum /jffs/$FILE | awk '{print $1}' 2>/dev/null)
+		MTD_MD5=$(md5sum /cifs2/$FILE | awk '{print $1}' 2>/dev/null)
+		if [ "${USB_MD5}" != "${MTD_MD5}" ];then
+			echo "更新文件：${FILE}"
+			cp -af /jffs/${FILE} /cifs2/${FILE}
+		fi
+	done
+
+	rm -rf /tmp/upload/jffs_folders.txt
+	rm -rf /tmp/upload/cifs2_folders.txt
+	rm -rf /tmp/upload/jffs_files.txt
+	rm -rf /tmp/upload/cifs2_files.txt
+	sync
+	echo "同步完成！"
+}
+
+set_sync_job(){
+	usb2jffs_sync_time=`nvram get usb2jffs_sync_time`
+	if [ "$usb2jffs_sync_time" -eq "0" ]; then
+		echo "删除插件定时同步任务..."
+		sed -i '/usb2jffs_sync/d' /var/spool/cron/crontabs/* >/dev/null 2>&1
+	elif [ "$usb2jffs_sync_time" -gt "0" ]; then
+		echo "设置每隔${usb2jffs_sync_time}分钟同步文件..."
+		cru a usb2jffs_sync "*/${usb2jffs_sync_time} * * * * sh /usr/sbin/usb2jffs.sh sync"
+	fi
+}
+
+case $1 in
+mount)
+	echo "========================= USB2JFFS - 手动挂载 ========================"
+	mount_jffs
+	echo "========================= USB2JFFS - 运行完成 ========================"
+	;;
+unmount)
+	echo "========================= USB2JFFS - 手动卸载 ========================"
+	unmount_jffs
+	echo "========================= USB2JFFS - 运行完成 ========================"
+	;;
+remove)
+	echo "========================= USB2JFFS - 手动删除 ========================"
+	remove_jffs
+	echo "========================= USB2JFFS - 运行完成 ========================"
+	;;
+sync)
+	echo "========================= USB2JFFS - 手动同步 ========================"
+	sync_usb_mtd
+	echo "========================= USB2JFFS - 运行完成 ========================"
+	;;
+setjob)
+	echo "========================= USB2JFFS - 定时同步 ========================"
+	set_sync_job
+	echo "========================= USB2JFFS - 运行完成 ========================"
+	;;
+*)
+	echo "用法:usb2jffs.sh [mount] [unmount] [remove] [sync]"
+	;;
+esac
diff --git a/release/src/router/rc/usb.c b/release/src/router/rc/usb.c
index 2c98230030..68a085d670 100644
--- a/release/src/router/rc/usb.c
+++ b/release/src/router/rc/usb.c
@@ -1831,7 +1831,7 @@ _dprintf("%s: stop_cloudsync.\n", __func__);
 
 	for (count = 0; count < 35; count++) {
 		sync();
-		ret = umount(mnt->mnt_dir);
+		ret = strcmp(mnt->mnt_dir,"/jffs") == 0 ? 0 : umount(mnt->mnt_dir);
 		if (!ret)
 			break;
 
@@ -1891,6 +1891,10 @@ _dprintf("%s: stop_cloudsync.\n", __func__);
 			// Only delete the directory if it was auto-created
 			rmdir(mnt->mnt_dir);
 		}
+		if (nvram_get_int("usb2jffs_enable")) {
+			_dprintf("run scripts: usb2jffs.sh unmount\n");
+			eval("usb2jffs.sh", "unmount");
+		}
 	}
 
 #if defined(RTCONFIG_APP_PREINSTALLED) || defined(RTCONFIG_APP_NETINSTALLED)
@@ -2200,6 +2204,15 @@ _dprintf("usb_path: 4. don't set %s.\n", tmp);
 		if(ret == MOUNT_VAL_RW)
 			test_of_var_files(mountpoint);
 
+		if (nvram_get_int("usb2jffs_enable")) {
+			_dprintf("run scripts: usb2jffs.sh mount\n");
+			eval("usb2jffs.sh", "mount");
+#ifdef RTCONFIG_SOFTCENTER
+			_dprintf("run scripts: jffsinit.sh\n");
+			eval("jffsinit.sh");
+#endif
+		}
+
 		if (nvram_get_int("usb_automount"))
 			run_nvscript("script_usbmount", mountpoint, 3);
 
@@ -4994,6 +5007,7 @@ static void start_diskformat(char *port_path)
 		char *tfat_cmd[] = { "mkfatfs", "-l", disk_label, "-v", devpath, NULL };
 		char *tntfs_cmd[] = {"mkntfs", "-F", "-L", disk_label, "-v", devpath, NULL };
 		char *thfsplus_cmd[] = { "newfs_hfs", "-v", disk_label, devpath, NULL };
+		char *ext4_cmd[] = { "mke2fs", "-t", "ext4", "-L", disk_label, "-T", "largefile", "-m", "1", "-pF", devpath, NULL };
 #endif
 		// format partition.
 		snprintf(devpath, sizeof(devpath), "/dev/%s", di->device);
@@ -5019,6 +5033,9 @@ static void start_diskformat(char *port_path)
 		else if(!strcmp(disk_system, "thfsplus") && nvram_match("usb_hfs_mod", "tuxera")) {
 			cmd = thfsplus_cmd;
 		}
+		else if(!strcmp(disk_system, "ext4")) {
+			cmd = ext4_cmd;
+		}
 		else
 #endif
 		{
diff --git a/release/src/router/shared/defaults.c b/release/src/router/shared/defaults.c
index bc18e61a4e..4b87e85b8e 100644
--- a/release/src/router/shared/defaults.c
+++ b/release/src/router/shared/defaults.c
@@ -4271,6 +4271,9 @@ struct nvram_tuple router_defaults[] = {
 #endif
 /* End of K3-specific settings */
 
+	{ "usb2jffs_enable", "0", CKN_STR1, CKN_TYPE_DEFAULT, CKN_ACC_LEVEL_DEFAULT, CKN_ENC_DEFAULT, 0 },
+	{ "usb2jffs_rsync", "0", CKN_STR1, CKN_TYPE_DEFAULT, CKN_ACC_LEVEL_DEFAULT, CKN_ENC_DEFAULT, 0 },
+	{ "usb2jffs_sync_time", "43200", CKN_STR5, CKN_TYPE_DEFAULT, CKN_ACC_LEVEL_DEFAULT, CKN_ENC_DEFAULT, 0 },
 	{ "aae_disable_force", "0", CKN_STR1, CKN_TYPE_DEFAULT, CKN_ACC_LEVEL_DEFAULT, CKN_ENC_DEFAULT, 0 },
 
 	{ 0, 0, 0, 0, 0, 0, 0 }
diff --git a/release/src/router/www/Tools_Usb2Jffs.asp b/release/src/router/www/Tools_Usb2Jffs.asp
new file mode 100644
index 0000000000..260d1208bf
--- /dev/null
+++ b/release/src/router/www/Tools_Usb2Jffs.asp
@@ -0,0 +1,445 @@
+﻿<!DOCTYPE html
+	PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml">
+<html xmlns:v>
+
+<head>
+	<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
+	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
+	<meta HTTP-EQUIV="Pragma" CONTENT="no-cache">
+	<meta HTTP-EQUIV="Expires" CONTENT="-1">
+	<link rel="shortcut icon" href="images/favicon.png">
+	<link rel="icon" href="images/favicon.png">
+	<title>
+		<#Web_Title#> - USB2JFFS
+	</title>
+	<link rel="stylesheet" type="text/css" href="index_style.css">
+	<link rel="stylesheet" type="text/css" href="form_style.css">
+	<link rel="stylesheet" type="text/css" href="/js/table/table.css">
+	<script type="text/javascript" src="/state.js"></script>
+	<script type="text/javascript" src="/general.js"></script>
+	<script type="text/javascript" src="/popup.js"></script>
+	<script type="text/javascript" src="/help.js"></script>
+	<script type="text/javascript" src="/tmhist.js"></script>
+	<script type="text/javascript" src="/tmmenu.js"></script>
+	<script type="text/javascript" src="/client_function.js"></script>
+	<script type="text/javascript" src="/js/jquery.js"></script>
+	<script type="text/javascript" src="/switcherplugin/jquery.iphone-switch.js"></script>
+	<script type="text/javascript" src="/js/table/table.js"></script>
+	<style>
+		.ks_btn {
+			border: 1px solid #222;
+			font-size:10pt;
+			color: #fff;
+			padding: 5px 5px 5px 5px;
+			border-radius: 5px 5px 5px 5px;
+			width:14%;
+			vertical-align: middle;
+			background: linear-gradient(to bottom, #003333  0%, #000000 100%);
+		}
+		.ks_btn:hover {
+			border: 1px solid #222;
+			font-size:10pt;
+			color: #fff;
+			padding: 5px 5px 5px 5px;
+			border-radius: 5px 5px 5px 5px;
+			width:14%;
+			vertical-align: middle;
+			background: linear-gradient(to bottom, #27c9c9  0%, #279fd9 100%);
+		}
+		.show-btn1, .show-btn2 {
+			font-size:10pt;
+			color: #fff;
+			padding: 10px 3.75px;
+			border-radius: 5px 5px 0px 0px;
+			width:8.42%;
+			border-left: 1px solid #67767d;
+			border-top: 1px solid #67767d;
+			border-right: 1px solid #67767d;
+			border-bottom: none;
+			background: #67767d;
+		}
+		.show-btn1:hover, .show-btn2:hover, .active {
+			border: 1px solid #2f3a3e;
+			background: #2f3a3e;
+		}
+		input[type=button]:focus {
+			outline: none;
+		}
+	</style>
+	<script>
+		var mounted;
+		var flag_total = 0;
+
+		function applyRule() {
+			showLoading();
+			document.form.submit();
+		}
+
+		function actions(flag) {
+			switch (flag) {
+				case 1:
+					FormActions("start_apply.htm", "toolscript", "usb2jffs.sh mount", "3");
+					break;
+				case 2:
+					FormActions("start_apply.htm", "toolscript", "usb2jffs.sh unmount", "3");
+					break;
+				case 3:
+					if(!confirm('注意：此操作会删除U盘里的.usb_jffs文件夹并卸载，其它文件不受影响！\n\n确定删除吗？')){
+						return false;
+					} else {
+						FormActions("start_apply.htm", "toolscript", "usb2jffs.sh remove", "3");
+					}
+					break;
+				case 4:
+					FormActions("start_apply.htm", "toolscript", "usb2jffs.sh sync", "3");
+					break;
+				case 5:
+					FormActions("start_apply.htm", "toolscript", "usb2jffs.sh setjob", "1");
+					break;
+				default:
+			}
+			applyRule();
+		}
+
+		function get_disks() {
+			require(['/require/modules/diskList.js'], function (diskList) {
+				var usb_path = [<% show_usb_path(); %>][0];
+				var html = '';
+				html += '<thead>'
+				html += '<tr>'
+				html += '<td colspan="8">磁盘分区列表</td>'
+				html += '</tr>'
+				html += '</thead>'
+				html += '<tr>'
+				html += '<th style="width:auto">端口</th>'
+				html += '<th style="width:auto">名称</th>'
+				html += '<th style="width:auto">大小</th>'
+				html += '<th style="width:auto">已用</th>'
+				html += '<th style="width:auto">权限</th>'
+				html += '<th style="width:auto">格式</th>'
+				html += '<th style="width:auto">挂载点</th>'
+				html += '<th style="width:auto">路径</th>'
+				html += '</tr>'
+				if (usb_path == "") {
+					E("_mount_status").innerHTML = "USB没有插入存储器";
+					$('.bt1').attr("disabled", "disabled");
+				} else {
+					usbDevicesList = diskList.list();
+					for (var i = 0; i < usbDevicesList.length; ++i) {
+						for (var j = 0; j < usbDevicesList[i].partition.length; ++j) {
+							//append options
+							if (usbDevicesList[i].partition[j].status != "unmounted") {
+								if (usbDevicesList[i].partition[j].format.indexOf("ext3") != -1
+								 || usbDevicesList[i].partition[j].format.indexOf("ext4") != -1
+								 || usbDevicesList[i].partition[j].format.indexOf("thfsplus") != -1)
+								{
+									flag_total = 1;
+								}
+							}
+							var totalsize = ((usbDevicesList[i].partition[j].size) / 1000000).toFixed(2);
+							var usedsize = ((usbDevicesList[i].partition[j].used) / 1000000).toFixed(2);
+							var usedpercent = (usedsize / totalsize * 100).toFixed(2) + " %";
+							var used = usedsize + " GB" + " (" + usedpercent + ")"
+							html += '<tr>'
+							html += '<td>' + usbDevicesList[i].usbPath + '</td>'
+							html += '<td>' + usbDevicesList[i].deviceName + '</td>'
+							html += '<td>' + totalsize + " GB" + '</td>'
+							html += '<td>' + used + '</td>'
+							html += '<td>' + usbDevicesList[i].partition[j].status + '</td>'
+							html += '<td>' + usbDevicesList[i].partition[j].format + '</td>'
+							html += '<td>' + usbDevicesList[i].partition[j].mountPoint + '</td>'
+							html += '<td>' + '/tmp/mnt/' + usbDevicesList[i].partition[j].partName + '</td>'
+							html += '</tr>'
+						}
+					}
+
+					if (flag_total == 0) {
+						E("_mount_status").innerHTML = "存储器格式不正确";
+						$('.bt1').attr("disabled", "disabled");
+					}
+				}
+				$('#disk_table').html(html);
+			});
+		}
+
+		function select_tablet(w) {
+			for (var i = 1; i <= 2; i++) {
+				$('.show-btn' + i).removeClass('active');
+				$('#tablet_' + i).hide();
+			}
+			$('.show-btn' + w).addClass('active');
+			$('#tablet_' + w).show();
+			if (document.form.usb2jffs_enable.value == 1) {
+				if (mounted == 1 && w == 1) {
+					$('.sbt0').hide();
+					$('.sbt1').show();
+				}
+				if (mounted == 0 && w == 1) {
+					$('.sbt0').show();
+					$('.sbt1').hide();
+				}
+			} else {
+				$('.tr1').hide();
+			}
+		}
+
+		function show_status() {
+			E("_jffs_status").innerHTML = mem_stats_arr[7];
+			if (usb2jffs_mount == 1) {
+				E("_mount_status").innerHTML = "已挂载";
+			} else if (flag_total == 1) {
+				E("_mount_status").innerHTML = "可挂载";
+			}
+		}
+
+		function init() {
+			show_menu();
+			update_sysinfo();
+			select_tablet(1);
+		}
+
+		function update_sysinfo(e) {
+			$.ajax({
+				url: '/ajax_sysinfo.asp',
+				dataType: 'script',
+				error: function (xhr) {
+					setTimeout("update_sysinfo();", 1000);
+				},
+				success: function (response) {
+					get_disks();
+					show_status();
+					setTimeout("update_sysinfo();", 3000);
+				}
+			});
+		}
+	</script>
+</head>
+
+<body onload="init();" onunLoad="return unload_body();">
+	<div id="TopBanner"></div>
+	<div id="Loading" class="popup_bg"></div>
+	<iframe name="hidden_frame" id="hidden_frame" src="" width="0" height="0" frameborder="0"></iframe>
+	<form method="post" name="form" id="ruleForm" action="/start_apply.htm" target="hidden_frame">
+		<input type="hidden" name="current_page" value="Tools_Usb2Jffs.asp" />
+		<input type="hidden" name="next_page" value="" />
+		<input type="hidden" name="group_id" value="" />
+		<input type="hidden" name="modified" value="0" />
+		<input type="hidden" name="action_mode" value="apply_new" />
+		<input type="hidden" name="action_script" value="" />
+		<input type="hidden" name="action_wait" value="3" />
+		<input type="hidden" name="first_time" value="" />
+		<input type="hidden" name="preferred_lang" id="preferred_lang" value="<% nvram_get("preferred_lang"); %>" />
+		<input type="hidden" name="firmver" value="<% nvram_get("firmver"); %>" />
+		<input type="hidden" name="usb2jffs_enable" value="<% nvram_get("usb2jffs_enable"); %>">
+		<input type="hidden" name="usb2jffs_rsync" value="<% nvram_get("usb2jffs_rsync"); %>">
+		<input type="hidden" name="usb2jffs_sync_time" value="<% nvram_get("usb2jffs_sync_time"); %>">
+
+		<table class="content" align="center" cellpadding="0" cellspacing="0">
+			<tr>
+				<td width="17">&nbsp;</td>
+				<td valign="top" width="202">
+					<div id="mainMenu"></div>
+					<div id="subMenu"></div>
+				</td>
+
+				<td valign="top">
+					<div id="tabMenu" class="submenuBlock"></div>
+					<!--===================================Beginning of Main Content===========================================-->
+					<table width="98%" border="0" align="left" cellpadding="0" cellspacing="0">
+						<tr>
+							<td align="left" valign="top">
+								<table width="760px" border="0" cellpadding="4" cellspacing="0" class="FormTitle" id="FormTitle">
+									<tr>
+										<td bgcolor="#4D595D" valign="top">
+											<div>&nbsp;</div>
+											<div class="formfonttitle">
+												<#Tools#> - USB2JFFS
+											</div>
+											<div style="margin:10px 0 10px 5px;" class="splitLine"></div>
+											<div class="formfontdesc">
+												轻松使用U盘挂载系统JFFS分区。
+											</div>
+											<div id="usb2jffs_disks_status" style="margin:10px 0px 0px 0px;">
+												<table id="disk_table" width="100%" border="1" align="center" cellpadding="4"
+													cellspacing="0" bordercolor="#6b8fa3" class="FormTable">
+													<thead>
+														<tr>
+															<td colspan="8">磁盘分区列表</td>
+														</tr>
+													</thead>
+													<tr>
+														<th style="width:auto">端口</th>
+														<th style="width:auto">名称</th>
+														<th style="width:auto">大小</th>
+														<th style="width:auto">已用</th>
+														<th style="width:auto">权限</th>
+														<th style="width:auto">格式</th>
+														<th style="width:auto">挂载点</th>
+														<th style="width:auto">路径</th>
+													</tr>
+												</table>
+											</div>
+											<div id="usb2jffs_mount_status" style="margin:10px 0px 0px 0px;">
+												<table id="status_pannel" width="100%" border="1" align="center" cellpadding="4" cellspacing="0" bordercolor="#6b8fa3" class="FormTable">
+													<thead>
+														<tr>
+															<td colspan="2">状态</td>
+														</tr>
+													</thead>
+													<tr>
+														<th><#JFFS#></th>
+														<td id="_jffs_status"><% sysinfo("jffs.usage"); %></td>
+													</tr>
+													<tr>
+														<th>挂载状态</th>
+														<td><span id="_mount_status">正在检测......</span></td>
+													</tr>
+												</table>
+											</div>
+
+											<div class="apply_gen">
+												<input style="cursor:pointer" type="button" class="button_gen" value="刷新" onclick="refreshpage();"/>
+											</div>
+
+											<div id="tablet_show">
+												<table style="margin:10px 0px 0px 0px;border-collapse:collapse" width="100%" height="37px">
+													<tr>
+														<td cellpadding="0" cellspacing="0" style="padding:0" border="1" bordercolor="#222">
+															<input id="show_btn1" class="show-btn1" style="cursor:pointer" type="button" value="操作" onclick="select_tablet(1);"/>
+															<input id="show_btn2" class="show-btn2" style="cursor:pointer" type="button" value="帮助" onclick="select_tablet(2);"/>
+														</td>
+													</tr>
+												</table>
+											</div>
+
+											<div id="tablet_1">
+												<table id="usb2jffs_config" width="100%" border="0" align="center" cellpadding="4" cellspacing="0" bordercolor="#6b8fa3" class="FormTable">
+													<tr>
+														<th>启用USB2JFFS</th>
+														<td>
+															<div align="center" class="left" style="width:94px; float:left; cursor:pointer;" id="usb2jffs_enable"></div>
+															<div class="iphone_switch_container" style="height:32px; width:74px; position: relative; overflow: hidden">
+																<script type="text/javascript">
+																	$('#usb2jffs_enable').iphoneSwitch('<% nvram_get("usb2jffs_enable"); %>',
+																		function(){
+																			document.form.usb2jffs_enable.value = 1;
+																			applyRule();
+																			$(".tr1").show();
+																			select_tablet(1);
+																		},
+																		function(){
+																			document.form.usb2jffs_enable.value = 0;
+																			actions(2);
+																			applyRule();
+																			$(".tr1").hide();
+																			select_tablet(1);
+																		}
+																	);
+																</script>
+															</div>
+														</td>
+													</tr>
+													<tr id="usb2jffs_sync_tr" class="tr1">
+														<th>定时同步（USB_JFFS → MTD_JFFS）</th>
+														<td>
+															每隔
+															<select name="usb2jffs_sync_time" class="input_option" style="width:auto">
+																<option value="0"
+																	<% nvram_match("usb2jffs_sync_time", "0","selected"); %>>不同步
+																</option>
+																<option value="3600"
+																	<% nvram_match("usb2jffs_sync_time", "3600","selected"); %>>1小时
+																</option>
+																<option value="7200"
+																	<% nvram_match("usb2jffs_sync_time", "7200","selected"); %>>2小时
+																</option>
+																<option value="14400"
+																	<% nvram_match("usb2jffs_sync_time", "14400","selected"); %>>4小时
+																</option>
+																<option value="21600"
+																	<% nvram_match("usb2jffs_sync_time", "21600","selected"); %>>6小时
+																</option>
+																<option value="43200"
+																	<% nvram_match("usb2jffs_sync_time", "43200","selected"); %>>12小时
+																</option>
+																<option value="86400"
+																	<% nvram_match("usb2jffs_sync_time", "86400","selected"); %>>24小时
+																</option>
+															</select>
+															<a type="button" class="ks_btn" style="cursor: pointer;margin-left:5px;border:none" onClick="actions(4);">手动同步</a>
+														</td>
+													</tr>
+													<tr id="usb2jffs_rsync_tr" class="tr1">
+														<th>卸载时同步</th>
+														<td>
+															<input type="radio" name="usb2jffs_rsync" class="input" value="1" <% nvram_match_x("", "usb2jffs_rsync", "1", "checked"); %>><#checkbox_Yes#>
+															<input type="radio" name="usb2jffs_rsync" class="input" value="0" <% nvram_match_x("", "usb2jffs_rsync", "0", "checked"); %>><#checkbox_No#>
+														</td>
+													</tr>
+													<tr id="operation_tr" class="tr1">
+														<th>操作</th>
+														<td>
+															<input class="ks_btn bt1" type="button" style="cursor: pointer;margin-left:5px;border:none" onClick="actions(5);" value="保存设置">
+															<input class="ks_btn bt1 sbt0" type="button" style="cursor: pointer;margin-left:5px;border:none" onClick="actions(1);" value="挂载">
+															<input class="ks_btn bt1 sbt1" type="button" style="cursor: pointer;margin-left:5px;border:none" onClick="actions(2);" value="卸载">
+															<input class="ks_btn bt1 sbt1" type="button" style="cursor: pointer;margin-left:5px;border:none" onClick="actions(3);" value="删除">
+														</td>
+													</tr>
+												</table>
+											</div>
+
+											<div id="tablet_2" style="display: none;">
+												<table style="margin:-1px 0px 0px 0px;" width="100%" border="1" align="center" cellpadding="4" cellspacing="0" bordercolor="#6b8fa3" class="FormTable">
+													<tr>
+													<td>
+														<ul>
+														<span style="line-height:2em"><b>功能说明</b></span><br />
+															&nbsp;&nbsp;&nbsp;&nbsp;1. 功能实现参考了：<a style="color:#17ffe4" href="https://koolshare.cn/thread-161017-1-1.html" title="" target="_blank">此koolshare论坛贴</a>，和软件中心的USB2JFFS插件。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;2. 打开启用开关后每次开机及插入USB时会自动尝试挂载。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;3. 卸载USB时会自动恢复原始的JFFS分区。<br />
+														<span style="line-height:2em"><b>使用说明</b></span><br />
+															&nbsp;&nbsp;&nbsp;&nbsp;1. 需要提前准备一个已经格式化成ext3/ext4/hfs格式的USB存储器。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;2. USB存储器有多个分区时只使用第一个符合要求的分区（按系统识别顺序）。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;3. 分区卷标不要使用中文，固件不能识别中文卷标。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;4. 使用的U盘质量不能太差，要求读写速度不能太低，可长时间稳定运行，不然可能极度影响使用体验。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;5. 建议挂载了USB JFFS后保持路由器长时间开机，不要频繁重启。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;6. 不要与其它USB2JFFS插件同时启用，以免造成混乱。<br />
+														<span style="line-height:2em"><b>操作说明</b></span><br />
+															&nbsp;&nbsp;&nbsp;&nbsp;1. 挂载： 插入U盘，点击挂载，插件会在USB磁盘中创建.usb_jffs目录，并用此目录挂载到/jffs。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;2. 卸载： 仅会移除/jffs与USB磁盘的挂载关系，并不会删除U盘和JFFS里的任何内容。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;3. 删除： 不仅会移除/jffs与USB磁盘的挂载关系，还会删除USB磁盘里的.usb_jffs目录。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;4. 同步： 点击同步，会将USB磁盘里.usb_jffs目录和路由器里的原始jffs挂载点分区进行同步。<br />
+														<span style="line-height:2em"><b>注意事项</b></span><br />
+															&nbsp;&nbsp;文件同步：<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;1. 首次成功挂载后，会复制原jffs目录下所有文件到USB的新jffs目录（即.usb_jffs文件夹）。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;2. 因路由器系统相关文件写入/更新、软件中心插件安装/更新等操作，USB的jffs内的文件会比原jffs新。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;3. 设定自动同步，或者手动点击同步按钮，会将新jffs下的文件和原jffs文件进行对比，并进行更新。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;4. 点击同步按钮不宜太频繁，避免文件频繁更新导致原jffs出现问题。<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;5. 如果安装了很多插件导致jffs目录大小大于原jffs分区大小，同步功能将只同步系统相关文件，如证书文件等。<br />
+															&nbsp;&nbsp;卸载USB存储器：<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;1. 请勿在已经挂载了USB型JFFS的状态下拔出USB存储器！<br />
+															&nbsp;&nbsp;&nbsp;&nbsp;2. 如果需要，请先在本插件点击卸载按钮，卸载USB型JFFS，再使用系统自带的移除USB硬盘，最后拔掉USB存储器。<br />
+														</ul>
+													</td>
+													</tr>
+												</table>
+											</div>
+										</td>
+									</tr>
+								</table>
+							</td>
+
+						</tr>
+					</table>
+					<!--===================================Ending of Main Content===========================================-->
+				</td>
+
+				<td width="10" align="center" valign="top">&nbsp;</td>
+			</tr>
+		</table>
+	</form>
+	<div id="footer"></div>
+</body>
+
+</html>
diff --git a/release/src/router/www/ajax_sysinfo.asp b/release/src/router/www/ajax_sysinfo.asp
index f9bdf45a87..558fb7c444 100644
--- a/release/src/router/www/ajax_sysinfo.asp
+++ b/release/src/router/www/ajax_sysinfo.asp
@@ -13,3 +13,5 @@ mem_stats_arr = ["<% sysinfo("memory.total"); %>",  "<% sysinfo("memory.free");
 	         "<% sysinfo("nvram.used"); %>", "<% sysinfo("jffs.usage"); %>"];
 
 cpu_stats_arr = ["<% sysinfo("cpu.load.1"); %>", "<% sysinfo("cpu.load.5"); %>", "<% sysinfo("cpu.load.15"); %>"];
+
+usb2jffs_mount = <% sysinfo("usb2jffs.mount"); %>;
diff --git a/release/src/router/www/device-map/disk_format.asp b/release/src/router/www/device-map/disk_format.asp
index 39c93f0c82..5c97dd7022 100644
--- a/release/src/router/www/device-map/disk_format.asp
+++ b/release/src/router/www/device-map/disk_format.asp
@@ -89,6 +89,10 @@ function set_disk_info(device) {
 			document.form.disk_name.maxLength = 30;
 			disk_system = "thfsplus";
 			break;
+		case "ext4" :
+			document.form.disk_name.maxLength = 16;
+			disk_system = "ext4";
+			break;
 		default :
 			document.form.disk_name.maxLength = 11;
 			disk_system = "tfat";
@@ -186,6 +190,10 @@ function go_format() {
 		if(temp_label.length > 31)
 			document.form.disk_name.value = temp_label.substr(0, 30);
 	}
+	else if(disk_system == "ext4") {
+		if(temp_label.length > 17)
+			document.form.disk_name.value = temp_label.substr(0, 16);
+	}
 
 	if(!Block_chars(document.form.disk_name, ["~", "`", "!", "#", "$", "%", "^", "&", "*", "(", ")", "+", "=", "{", "[", "}", "]", "|", "\\", ":", ";", "\"", "'", "<", ">", ",", ".", "?", "/", " "]))
 		return false;
@@ -224,6 +232,9 @@ function change_disk_system() {
 		if(temp_label.length > 31)
 			document.form.disk_name.value = temp_label.substr(0, 30);
 	}
+	else if(disk_system == "ext4") {
+		document.form.disk_name.maxLength = 16;
+	}
 }
 function show_loadingBar_field(){
 	document.getElementById('loadingIcon').style.display = "none";
@@ -366,7 +377,7 @@ function showLoadingUpdate(){
 			<div>
 				<input type="text" maxlength="32" class="input_25_table" name="disk_name" value="" onkeypress="return validator.isString(this, event)" autocorrect="off" autocapitalize="off">
 			</div>
-			
+
 		</div>
 		<div style="margin-top:5px;*margin-top:-10px;width:283px;" class="line_horizontal"></div>
 		<div style="padding:7px;">
@@ -376,6 +387,7 @@ function showLoadingUpdate(){
 					<option value="tntfs">NTFS</option>
 					<option value="tfat">FAT</option>
 					<option value="thfsplus">HFS</option>
+					<option value="ext4">EXT4</option>
 				</select>
 			</div>
 		</div>
diff --git a/release/src/router/www/require/menuTrees/menuTree_bwdpi_traffic_analyzer.js b/release/src/router/www/require/menuTrees/menuTree_bwdpi_traffic_analyzer.js
index cd5e21bef8..ad181e5cc0 100644
--- a/release/src/router/www/require/menuTrees/menuTree_bwdpi_traffic_analyzer.js
+++ b/release/src/router/www/require/menuTrees/menuTree_bwdpi_traffic_analyzer.js
@@ -203,6 +203,7 @@ define(function(){
 					{url: "Tools_Sysinfo.asp", tabName: "<#Sysinfo#>"},
 					{url: "Tools_OtherSettings.asp", tabName: "<#Other_Settings#>"},
 					{url: "Tools_K3Screen.asp", tabName: "<#K3Screen#>"},
+					{url: "Tools_Usb2Jffs.asp", tabName: "USB2JFFS"},
 					{url: "NULL", tabName: "__INHERIT__"}
 				]
 			},
diff --git a/release/src/router/www/require/menuTrees/menuTree_no_bwdpi.js b/release/src/router/www/require/menuTrees/menuTree_no_bwdpi.js
index bec99ba21d..ead83e8cf7 100644
--- a/release/src/router/www/require/menuTrees/menuTree_no_bwdpi.js
+++ b/release/src/router/www/require/menuTrees/menuTree_no_bwdpi.js
@@ -150,6 +150,7 @@ define(function(){
 					{url: "Tools_Sysinfo.asp", tabName: "<#Sysinfo#>"},
 					{url: "Tools_OtherSettings.asp", tabName: "<#Other_Settings#>"},
 					{url: "Tools_K3Screen.asp", tabName: "<#K3Screen#>"},
+					{url: "Tools_Usb2Jffs.asp", tabName: "USB2JFFS"},
 					{url: "NULL", tabName: "__INHERIT__"}
 				]
 			},
-- 
2.25.1

